File: src/registry.py
from abc import ABC, abstractmethod
from typing import List, Optional


class Registry(ABC):
    @abstractmethod
    def register(self, url: str, port: Optional[int] = None, **kwargs):
        """
        Register a backend service. Accepts URL, port, and optional metadata.
        """
        pass

    @abstractmethod
    def unregister(self, url: str, port: Optional[int] = None, **kwargs):
        """
        Unregister a backend service. Accepts URL, port, and optional metadata.
        """
        pass

    @abstractmethod
    def list_backends(self) -> List[dict]:
        """
        Return a list of all registered backends and their metadata.
        """
        pass

File: go/abstractions/registry.go
package abstractions

import (
	"openprequal/contracts"
	"sync"
)

// Registry is an interface for registering and listing backends.
type Registry interface {
	Register(b contracts.Backend)
	Unregister(b contracts.Backend)
	ListBackends() []contracts.Backend
}

// NewInMemoryRegistry creates a simple in-memory registry.
func NewInMemoryRegistry() *InMemoryRegistry {
	return &InMemoryRegistry{backends: make([]contracts.Backend, 0)}
}

// InMemoryRegistry is a thread-safe in-memory implementation of Registry.
type InMemoryRegistry struct {
	mu       sync.Mutex
	backends []contracts.Backend
}

// Register adds a backend to the registry.
func (r *InMemoryRegistry) Register(b contracts.Backend) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.backends = append(r.backends, b)
}

// Unregister removes backends that match url and port.
func (r *InMemoryRegistry) Unregister(b contracts.Backend) {
	r.mu.Lock()
	defer r.mu.Unlock()
	out := r.backends[:0]
	for _, be := range r.backends {
		if be.Equal(&b) {
			continue
		}
		out = append(out, be)
	}
	r.backends = out
}

// ListBackends returns a copy of backends.
func (r *InMemoryRegistry) ListBackends() []contracts.Backend {
	r.mu.Lock()
	defer r.mu.Unlock()
	copyOut := make([]contracts.Backend, len(r.backends))
	copy(copyOut, r.backends)
	return copyOut
}

// ModelDump returns a map representation of the backend (rough parity with Python model_dump())
// ModelDump now lives on contracts.Backend
